src/molecules/lists/drag-and-drop-list-box.tsx

```diff
@@ -0,0 +1,212 @@
+// #region Interfaces
+// -------------------------------------------------------------------------------------------------
+
+export interface DragAndDropListBoxProps<T extends any>
+    extends Omit<ListBoxProps<T>, "children" | "items"> {
+    droppableId: string;
+    items: Array<ListBoxItem<T>>;
+    onReordered: (newOrderedArray: Array<ListBoxItem<T>>) => void;
+}
+
+// #endregion Interfaces
+
+// -------------------------------------------------------------------------------------------------
+// #region Component
+// -------------------------------------------------------------------------------------------------
+
+const DragAndDropListBox = <T extends any>(
+    props: DragAndDropListBoxProps<T>
+): ReactElement<ListBoxProps<T>> | null => {
+    const items = props.items;
+    const hasItems = CollectionUtils.hasValues(items);
+
+    // Short-circuit if no items in collection
+    if (items != null && !hasItems) {
+        if (props.hideWhenNoItems) {
+            return null;
+        }
+
+        return (
+            <div className={`${COMPONENT_CLASS} -empty`}>
+                <Paragraph>No Items Added</Paragraph>
+            </div>
+        );
+    }
+
+    const processReordered = (
+        items: Array<ListBoxItem<T>>,
+        startIndex: number,
+        endIndex: number
+    ): Array<ListBoxItem<T>> => {
+        const result = [...items];
+        const [removedItem] = result.splice(startIndex, 1);
+        result.splice(endIndex, 0, removedItem);
+
+        return result;
+    };
+
+    const handleDragEnd = (result: DropResult) => {
+        if (
+            result?.destination == null ||
+            result?.source == null ||
+            result.source.index === result.destination.index
+        ) {
+            return;
+        }
+
+        props.onReordered(
+            processReordered(
+                props.items,
+                result.source.index,
+                result.destination!.index
+            )
+        );
+    };
+
+    /**
+     * if we don't memoize the child rendering, then the <Draggable> state
+     * sometimes does not update properly. Without memoization, if you drag
+     * and drop a <Draggable> outside it's <Droppable>, then the <Draggable> will
+     * still think it's being dragged, and apply the -dragging class, even though
+     * it is no longer being dragged. Memoization of rendering the list fixes
+     * this issue, and also increases performance.
+     */
+    const DraggableItemList = React.memo(function(itemProps: {
+        items: Array<ListBoxItem<T>>;
+    }) {
+        const cssIsDragging = (snapshot: DraggableStateSnapshot) =>
+            snapshot.isDragging ? "-dragging" : "";
+
+        return (
+            <React.Fragment>
+                {itemProps.items.map((item: ListBoxItem<T>, index: number) => (
+                    <Draggable
+                        draggableId={`${item.id}`}

```

Comment:
Nitpick - if you have some extra time to dig into it, I'd love to get the typing correct so we could just call `toString()` on the item id instead of using string interpolation on the whole thing.